# Python For Data Engineering part 2 – the TD

## The CSV analysis

download the mobike csv usign baidu pan

链接: https://pan.baidu.com/s/1aszR_8xX9cvniCeGO9OwaA  

密码: kak5
--来自百度网盘超级会员V3的分享

Using the mobike dataset, read the csv and answer the following questions:

1. Which day of the week see the most events?
2. Which day of the month see the most events? 
3. Which bike has been used the most?
4. How many bike does a user use in average?
5. Which day of the week has the longest average travel?

In [5]:
import csv
import datetime
import requests
with open('UTSEUS-MOBIKE-shanghai_full.csv',"r") as mobike:
    csvreader = csv.reader(mobike)
    mobikedata = [row for i, row in enumerate(mobike)]

<font size=5>Q1:Which day of the week see the most events?</font>

In [2]:
#q1
date_dict={1:0,2:0,3:0,4:0,5:0,6:0,7:0}
with open('UTSEUS-MOBIKE-shanghai_full.csv',"r") as mobike:
    csvreader = csv.reader(mobike)
    next(mobike)
    for row in csvreader:
        test_row=row[3].strip()
        weekday=datetime.date(int(test_row[0:4]),int(test_row[5:7]),int(test_row[8:10])).isoweekday()
        date_dict[weekday]+= 1
date_dict

{1: 162692, 2: 161838, 3: 177561, 4: 125751, 5: 132976, 6: 133356, 7: 129429}

Wedneseday see the most events

<font size=5>Q2:Which day of the month see the most events?</font>

In [3]:
#q2
day=[]
with open('UTSEUS-MOBIKE-shanghai_full.csv','r') as mobike:
    csvreader=csv.reader(mobike)
    next(mobike)
    for row in csvreader:
        test_row=row[3].strip()
        day.append(int(test_row[8:10]))
time_day=[]
for i in range(1,32):
    time_day.append(day.count(i))
time_day.index(max(time_day))+1

31

the 31th day of the month see the most events

<font size=5>Q3:Which bike has been used the most?</font>

In [2]:
#q3
bike=[]
times=[]
with open('UTSEUS-MOBIKE-shanghai_full.csv','r') as mobike:
    csvreader=csv.reader(mobike)
    next(mobike)
    for row in csvreader:
        test_row=row[1].strip()
        if(test_row in bike):
            times[bike.index(test_row)]+=1
        else:
            bike.append(test_row)
            times.append(1)
max_time=max(times)
index_max_time=times.index(max_time)
print(bike[index_max_time])

32253


The bike whose id is '32253' has benn used the most

<font size=5>Q4:How many bike does a user use in average?</font>

In [6]:
#q4
user_dict=[]
num_bike=0
with open('UTSEUS-MOBIKE-shanghai_full.csv','r') as mobike:
    csvreader=csv.reader(mobike)
    next(mobike)
    for row in csvreader:
        test_row=row[2].strip()
        num_bike+=1
        if(test_row not in user_dict):
            user_dict.append(test_row)
average=round(num_bike/len(user_dict),2)
print(average)

57.87


a user uses 57.87 bikes in average 

<font size=5>Q5:Which day of the week has the longest average travel?</font>

In [8]:
#q5
sum_dict={1:0,2:0,3:0,4:0,5:0,6:0,7:0}
date_dict={1:0,2:0,3:0,4:0,5:0,6:0,7:0}
average_dict={1:0,2:0,3:0,4:0,5:0,6:0,7:0}
with open('UTSEUS-MOBIKE-shanghai_full.csv','r') as mobike:
    csvreader=csv.reader(mobike)
    next(mobike)
    for row in csvreader:
        start_row = row[3].strip()
        end_row = row[6].strip()
        start = int(start_row[-5:-3]) * 60 + int(start_row[-2:])
        end = int(end_row[-5:-3]) * 60 + int(end_row[-2:])
        difference = end - start
        time=datetime.date(int(start_row[0:4]),int(start_row[5:7]),int(start_row[8:10]))
        weekday=time.isoweekday()
        sum_dict[weekday]+=difference
        date_dict[weekday]+=1
for i in range(1,8):
    average_dict[i]=round(sum_dict[i]/date_dict[i],2)
print(average_dict)

{1: 9.16, 2: 9.6, 3: 10.17, 4: 10.21, 5: 5.84, 6: 6.85, 7: 10.42}


Sunday has the longest average travel

<font size=5>Challenge of API</font>

API

In [ ]:
isbn = '9782070754922'
url= f'https://openlibrary.org/api/books'
params = {'bibkeys':f'ISBN:{isbn}',
          'format':'json',
          'jscmd':'data'}
r=requests.get(url, params=params)
r.status_code

In [18]:
import requests

In [27]:
urlpara={
    'wd':'iphone',
    'rsv_spt':'1'
}
url='https://www.baidu.com/s'
r=requests.get(url,urlpara)
r.status_code

200

In [45]:
import requests
url_git='https://api.github.com/users/fpfaende'
r_git=requests.get(url_git)
r_git.status_code

200

In [46]:
response=r_git.json()
response['name']

'Fabien Pfaender'